jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -376,65 +386,41 @@ private void clearArray(RetainedBucket[] poolArray, AtomicLong memoryCounter)
     private void releaseExcessMemory(boolean direct)
     {
         long maxMemory = direct ? _maxDirectMemory : _maxHeapMemory;
-        if (maxMemory > 0)
-        {
-            long excess = getMemory(direct) - maxMemory;
-            if (excess > 0)
-                evict(direct, excess);
-        }
-    }
+        if (maxMemory <= 0)
+            return;
 
-    /**
-     * This eviction mechanism searches for the RetainableByteBuffers that were released the longest time ago.
-     *
-     * @param direct true to search in the direct buffers buckets, false to search in the heap buffers buckets.
-     * @param excess the amount of bytes to evict. At least this much will be removed from the buckets.
-     */
-    private void evict(boolean direct, long excess)
-    {
-        if (LOG.isDebugEnabled())
-            LOG.debug("evicting {} bytes from {} pools", excess, (direct ? "direct" : "heap"));
-        long now = NanoTime.now();
-        long totalClearedCapacity = 0L;
+        if (_evictor.compareAndSet(false, true))
+        {
+            // we are the one and only thread evicting
+            try
+            {
+                RetainedBucket[] buckets = direct ? _direct : _indirect;
 
-        RetainedBucket[] buckets = direct ? _direct : _indirect;
+                long excess = getMemory(direct) - maxMemory;
+                while (excess > 0)
```

Comment:
Don't like the loop.
@lorban has found that looping may not exit if the pool is under stress, as it would never find enough memory to return.

Reply:
@sbordet My loop is limited, as I only calculate the excess once, so it will never loop forever.
Let me look at your PR again, but last I looked it was evicting the wrong buffer... i.e the just released one rather than one that has been idle for some time.

Reply:
@gregw given that the strategy for the pool is THREAD_ID, all buffers are typically equally old, as they are picked randomly. So looking for the oldest is not optimal. At that point, any buffer to evict is a good buffer, and the current one is as good as the others.

Your loop may run forever if `excess` is never modified, which may be the case as evict() may return 0.

And does not take into account that a call to `entry.remove()` returning `false` will _eventually_ remove the entry, so your loop over the entries is removing _all_ the entries, basically emptying the bucket!

Reply:
@sbordet
> @gregw given that the strategy for the pool is THREAD_ID, all buffers are typically equally old, as they are picked randomly. So looking for the oldest is not optimal. At that point, any buffer to evict is a good buffer, and the current one is as good as the others.

The algorithm is not looking for the oldest buffer.  It is looking for the bucket with the most idle space.   It is so it can adapt to changing size demand.

> Your loop may run forever if `excess` is never modified, which may be the case as evict() may return 0.

Good point.   I'll make sure it is always limited

> And does not take into account that a call to `entry.remove()` returning `false` will _eventually_ remove the entry, so your loop over the entries is removing _all_ the entries, basically emptying the bucket!
 
Firstly it is always iterating over only idle buffers, so it can only ever remove all the idle entries of the bucket, which may not be a terrible algorithm.   But then remove is only going to return false if the entry has already been removed or if it is multiplexed.  The later doesn't apply, so false not a problem here.

Reply:
I've updated the eviction so that only a single pass is every done by one thread.  If that is not enough, it sets an atomic so the next releaser will try another pass.